src/qibo/models/iqae.py

```diff
@@ -0,0 +1,385 @@
+            k.append(k_i)
+            if K_i > int(L_max / eps):
+                N_shots_i = int((L_max / eps) * self.N_shots / K_i / 10)
+                if N_shots_i == 0:
+                    N_shots_i = 1
+            else:
+                N_shots_i = self.N_shots
+
+            # Calling the quantum circuit
+            qc = self.construct_circuit(k_i)
+            samples = self.run_circuit(qc, N_shots_i)
+
+            samples_history.append(samples)
+            n_shots_history.append(N_shots_i)
+
+            num_oracle_queries += N_shots_i * k_i
+
+            l = 1
+            if i > 1:
+                while K[i - l] == K[i] and i >= l + 1:
+                    l = l + 1
+                sum_total_samples = sum([samples_history[j] for j in range(i - l, i)])
+                N_total_shots = sum([n_shots_history[j] for j in range(i - l, i)])
+
+            else:
+                sum_total_samples = samples
+                N_total_shots = N_shots_i
+
+            a = sum_total_samples / N_total_shots
+
+            if self.method == "chernoff":
+                delta_a = np.sqrt(np.log(2 * T / self.alpha) / 2 / N_total_shots)
+                a_min_i = max(0, a - delta_a)
+                a_max_i = min(1, a + delta_a)
+            else:
+                a_min_i, a_max_i = self.clopper_pearson(
+                    sum_total_samples, N_total_shots, alpha=self.alpha / T
+                )
+            a_min.append(a_min_i)
+            a_max.append(a_max_i)
+
+            if up_i:
+                theta_min_i = np.arccos(1 - 2 * a_min_i) / 2 / np.pi
+                theta_max_i = np.arccos(1 - 2 * a_max_i) / 2 / np.pi
+            else:
+                theta_min_i = 1 - np.arccos(1 - 2 * a_max_i) / 2 / np.pi
+                theta_max_i = 1 - np.arccos(1 - 2 * a_min_i) / 2 / np.pi
+
+            theta = min(int(K_i * theta_u), int(K_i * theta_l))
+
+            theta_u = (theta + theta_max_i) / K_i
+            theta_l = (theta + theta_min_i) / K_i
+            theta_dif = np.abs(theta_u - theta_l)
+            theta_intervals.append([theta_l, theta_u])
+
+            a_l = np.sin(2 * np.pi * theta_l) ** 2
+            a_u = np.sin(2 * np.pi * theta_u) ** 2
+            a_intervals.append([a_l, a_u])
+
+        result = IterativeAmplitudeEstimationResult()
+        result.alpha = self.alpha
+        result.epsilon_target = self.epsilon
+        result.epsilon_estimated = (a_u - a_l) / 2
+        result.estimate_intervals = a_intervals
+        result.num_oracle_queries = num_oracle_queries
+        result.estimation = (a_u + a_l) / 2
+
+        result.theta_intervals = theta_intervals
+        result.k_list = k
+        result.ratios = samples_history
+        result.shots = n_shots_history
+
+        return result
+
+
+# THE CLASS WITH THE RESULTS
+class IterativeAmplitudeEstimationResult:
+    """The ``IterativeAmplitudeEstimation`` result object."""
+
+    def __init__(self):
+        super().__init__()
```

Comment:
`IterativeAmplitudeEstimationResult` is not a derived class, you can drop this line
```suggestion

```
If you want you can use [dataclasses](https://docs.python.org/3/library/dataclasses.html)